File: app/src/main/java/com/pengxh/autodingding/ui/HistoryRecordActivity.java
package com.pengxh.autodingding.ui;

import android.os.CountDownTimer;
import android.os.Environment;
import android.os.Handler;
import android.text.TextUtils;
import android.view.View;

import androidx.recyclerview.widget.LinearLayoutManager;

import com.gyf.immersionbar.ImmersionBar;
import com.pengxh.androidx.lite.base.AndroidxBaseActivity;
import com.pengxh.androidx.lite.utils.ColorUtil;
import com.pengxh.androidx.lite.utils.DeviceSizeUtil;
import com.pengxh.androidx.lite.utils.ImmerseStatusBarUtil;
import com.pengxh.androidx.lite.utils.SaveKeyValues;
import com.pengxh.androidx.lite.utils.WeakReferenceHandler;
import com.pengxh.androidx.lite.widget.EasyPopupWindow;
import com.pengxh.androidx.lite.widget.EasyToast;
import com.pengxh.androidx.lite.widget.dialog.AlertControlDialog;
import com.pengxh.androidx.lite.widget.dialog.AlertMessageDialog;
import com.pengxh.autodingding.BaseApplication;
import com.pengxh.autodingding.R;
import com.pengxh.autodingding.adapter.HistoryRecordAdapter;
import com.pengxh.autodingding.bean.HistoryRecordBean;
import com.pengxh.autodingding.databinding.ActivityHistoryBinding;
import com.pengxh.autodingding.greendao.DaoSession;
import com.pengxh.autodingding.utils.Constant;
import com.pengxh.autodingding.utils.ExcelUtils;

import java.io.File;
import java.util.ArrayList;
import java.util.List;

public class HistoryRecordActivity extends AndroidxBaseActivity<ActivityHistoryBinding> {

    private static final String TAG = "HistoryRecordActivity";
    private static final int[] images = new int[]{R.drawable.ic_delete, R.drawable.ic_export};
    private static final String[] titles = new String[]{"删除记录", "导出记录"};
    private static final String[] excelTitle = {"uuid", "日期", "打卡信息"};
    private EasyPopupWindow easyPopupWindow;
    private WeakReferenceHandler weakReferenceHandler;
    private DaoSession daoSession;
    private List<HistoryRecordBean> dataBeans = new ArrayList<>();
    private boolean isRefresh = false;
    private boolean isLoadMore = false;
    private int offset = 0;// 本地数据库分页从0开始
    private HistoryRecordAdapter historyAdapter;


    @Override
    protected void setupTopBarLayout() {
        ImmerseStatusBarUtil.setColor(this, ColorUtil.convertColor(this, R.color.colorAppThemeLight));
        ImmersionBar.with(this).statusBarDarkFont(false).init();
        viewBinding.titleView.setText("打卡记录");
        viewBinding.titleRightView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                easyPopupWindow.showAsDropDown(viewBinding.titleRightView
                        , 0
                        , DeviceSizeUtil.dp2px(HistoryRecordActivity.this, 10));
            }
        });
    }

    @Override
    public void initData() {
        weakReferenceHandler = new WeakReferenceHandler(callback);
        daoSession = BaseApplication.getInstance().getDaoSession();

        dataBeans = queryHistoryRecord();
        weakReferenceHandler.sendEmptyMessage(2022021403);

        easyPopupWindow = new EasyPopupWindow(this);
        easyPopupWindow.setPopupMenuItem(images, titles);
        easyPopupWindow.setOnPopupWindowClickListener(position -> {
            if (position == 0) {
                //添加导出功能
                if (dataBeans.size() == 0) {
                    new AlertMessageDialog.Builder()
                            .setContext(this)
                            .setTitle("温馨提示")
                            .setMessage("空空如也，无法删除")
                            .setPositiveButton("确定")
                            .setOnDialogButtonClickListener(() -> {

                            }).build().show();
                } else {
                    new AlertControlDialog.Builder()
                            .setContext(this)
                            .setTitle("温馨提示")
                            .setMessage("是否确定清除打卡记录？")
                            .setNegativeButton("取消")
                            .setPositiveButton("确定")
                            .setOnDialogButtonClickListener(new AlertControlDialog.OnDialogButtonClickListener() {
                                @Override
                                public void onConfirmClick() {
                                    daoSession.getHistoryRecordBeanDao().deleteAll();
                                    dataBeans.clear();
                                    historyAdapter.notifyDataSetChanged();
                                    viewBinding.emptyView.setVisibility(View.VISIBLE);
                                }

                                @Override
                                public void onCancelClick() {

                                }
                            }).build().show();
                }
            } else if (position == 1) {
                String emailAddress = (String) SaveKeyValues.getValue(Constant.EMAIL_ADDRESS, "");
                if (TextUtils.isEmpty(emailAddress)) {
                    EasyToast.show(this, "未设置邮箱，无法导出");
                    return;
                }
                if (dataBeans.size() == 0) {
                    EasyToast.show(this, "无打卡记录，无法导出");
                    return;
                }
                new AlertControlDialog.Builder()
                        .setContext(this)
                        .setTitle("温馨提示")
                        .setMessage("导出到" + emailAddress + "？")
                        .setNegativeButton("取消")
                        .setPositiveButton("确定")
                        .setOnDialogButtonClickListener(new AlertControlDialog.OnDialogButtonClickListener() {
                            @Override
                            public void onConfirmClick() {
                                //导出Excel
                                exportToEmail(dataBeans);
                            }

                            @Override
                            public void onCancelClick() {

                            }
                        }).build().show();
            }
        });
    }

    @Override
    public void initEvent() {
        viewBinding.refreshLayout.setOnRefreshListener(refreshLayout -> {
            isRefresh = true;
            new CountDownTimer(1000, 500) {

                @Override
                public void onTick(long millisUntilFinished) {

                }

                @Override
                public void onFinish() {
                    isRefresh = false;
                    dataBeans.clear();
                    offset = 0;
                    dataBeans = queryHistoryRecord();

                    refreshLayout.finishRefresh();
                    weakReferenceHandler.sendEmptyMessage(2022021403);
                }
            }.start();
        });

        viewBinding.refreshLayout.setOnLoadMoreListener(refreshLayout -> {
            isLoadMore = true;
            new CountDownTimer(1000, 500) {

                @Override
                public void onTick(long millisUntilFinished) {

                }

                @Override
                public void onFinish() {
                    isLoadMore = false;
                    offset++;
                    dataBeans.addAll(queryHistoryRecord());

                    refreshLayout.finishLoadMore();
                    weakReferenceHandler.sendEmptyMessage(2022021403);
                }
            }.start();
        });
    }

    private final Handler.Callback callback = msg -> {
        if (msg.what == 2022021403) {
            if (isRefresh || isLoadMore) {
                historyAdapter.notifyDataSetChanged();
            } else { //首次加载数据
                if (dataBeans.size() == 0) {
                    viewBinding.emptyView.setVisibility(View.VISIBLE);
                } else {
                    viewBinding.emptyView.setVisibility(View.GONE);
                    historyAdapter = new HistoryRecordAdapter(this, dataBeans);
                    viewBinding.historyRecordView.setLayoutManager(new LinearLayoutManager(this));
                    viewBinding.historyRecordView.setAdapter(historyAdapter);
                }
            }
        }
        return true;
    };

    private List<HistoryRecordBean> queryHistoryRecord() {
        return daoSession.queryBuilder(HistoryRecordBean.class).offset(offset * 15).limit(15).list();
    }

    private void exportToEmail(List<HistoryRecordBean> historyBeans) {
        //{"date":"2020-04-15","message":"考勤打卡:11:42 下班打卡 早退","uuid":"26btND0uLqU"},{"date":"2020-04-15","message":"考勤打卡:16:32 下班打卡 早退","uuid":"UTWQJzCfTl9"}
        File dir = new File(this.getExternalFilesDir(Environment.DIRECTORY_DOCUMENTS), "DingRecord");
        if (!dir.exists()) {
            dir.mkdir();
        }
        ExcelUtils.initExcel(dir + "/打卡记录表.xls", excelTitle);
        String fileName = dir + "/打卡记录表.xls";
        ExcelUtils.writeObjListToExcel(this, historyBeans, fileName);
    }
}

File: app/src/main/java/com/pengxh/autodingding/ui/HistoryRecordActivity.kt
package com.pengxh.autodingding.ui

import android.content.Context
import android.graphics.Color
import android.os.Bundle
import android.os.CountDownTimer
import android.os.Environment
import android.os.Handler
import android.os.Message
import android.text.TextUtils
import android.view.View
import com.pengxh.autodingding.BaseApplication
import com.pengxh.autodingding.R
import com.pengxh.autodingding.bean.HistoryRecordBean
import com.pengxh.autodingding.databinding.ActivityHistoryBinding
import com.pengxh.autodingding.extensions.initImmersionBar
import com.pengxh.autodingding.extensions.writeObjToExcel
import com.pengxh.autodingding.greendao.HistoryRecordBeanDao
import com.pengxh.autodingding.utils.Constant
import com.pengxh.autodingding.utils.ExcelUtils
import com.pengxh.kt.lite.adapter.NormalRecyclerAdapter
import com.pengxh.kt.lite.adapter.ViewHolder
import com.pengxh.kt.lite.base.KotlinBaseActivity
import com.pengxh.kt.lite.divider.RecyclerViewItemDivider
import com.pengxh.kt.lite.extensions.dp2px
import com.pengxh.kt.lite.extensions.show
import com.pengxh.kt.lite.utils.SaveKeyValues
import com.pengxh.kt.lite.utils.WeakReferenceHandler
import com.pengxh.kt.lite.widget.EasyPopupWindow
import com.pengxh.kt.lite.widget.dialog.AlertControlDialog
import com.pengxh.kt.lite.widget.dialog.AlertMessageDialog
import java.io.File

class HistoryRecordActivity : KotlinBaseActivity<ActivityHistoryBinding>() {

    private val context: Context = this@HistoryRecordActivity
    private val images = intArrayOf(R.drawable.ic_delete, R.drawable.ic_export)
    private val titles = arrayOf("删除记录", "导出记录")
    private val excelTitle = arrayOf("uuid", "日期", "打卡信息")
    private val historyRecordBeanDao by lazy { BaseApplication.get().daoSession.historyRecordBeanDao }
    private val easyPopupWindow by lazy { EasyPopupWindow(this) }
    private lateinit var weakReferenceHandler: WeakReferenceHandler
    private lateinit var historyAdapter: NormalRecyclerAdapter<HistoryRecordBean>
    private var dataBeans: MutableList<HistoryRecordBean> = ArrayList()
    private var isRefresh = false
    private var isLoadMore = false
    private var offset = 0 // 本地数据库分页从0开始

    override fun initViewBinding(): ActivityHistoryBinding {
        return ActivityHistoryBinding.inflate(layoutInflater)
    }

    override fun setupTopBarLayout() {
        binding.rootView.initImmersionBar(this, false, R.color.colorAppThemeLight)
        binding.titleView.text = "打卡记录"
        binding.titleRightView.setOnClickListener {
            easyPopupWindow.showAsDropDown(binding.titleRightView, 0, 10.dp2px(context))
        }
    }

    override fun initOnCreate(savedInstanceState: Bundle?) {
        weakReferenceHandler = WeakReferenceHandler(callback)
        dataBeans = queryHistoryRecord()
        weakReferenceHandler.sendEmptyMessage(2022021403)

        val items = ArrayList<EasyPopupWindow.MenuItem>()
        items.add(EasyPopupWindow.MenuItem(images[0], titles[0]))
        items.add(EasyPopupWindow.MenuItem(images[1], titles[1]))
        easyPopupWindow.set(items, object : EasyPopupWindow.OnPopupWindowClickListener {
            override fun onPopupItemClicked(position: Int) {
                when (position) {
                    0 -> {
                        if (dataBeans.size == 0) {
                            AlertMessageDialog.Builder()
                                .setContext(context)
                                .setTitle("温馨提示")
                                .setMessage("空空如也，无法删除")
                                .setPositiveButton("确定")
                                .setOnDialogButtonClickListener(
                                    object : AlertMessageDialog.OnDialogButtonClickListener {
                                        override fun onConfirmClick() {

                                        }
                                    }
                                ).build().show()
                        } else {
                            AlertControlDialog.Builder()
                                .setContext(context)
                                .setTitle("温馨提示")
                                .setMessage("是否确定清除打卡记录？")
                                .setNegativeButton("取消")
                                .setPositiveButton("确定")
                                .setOnDialogButtonClickListener(object :
                                    AlertControlDialog.OnDialogButtonClickListener {
                                    override fun onConfirmClick() {
                                        historyRecordBeanDao.deleteAll()
                                        dataBeans.clear()
                                        historyAdapter.notifyDataSetChanged()
                                        binding.emptyView.visibility = View.VISIBLE
                                    }

                                    override fun onCancelClick() {

                                    }
                                }).build().show()
                        }
                    }

                    1 -> {
                        val emailAddress =
                            SaveKeyValues.getValue(Constant.EMAIL_ADDRESS, "") as String
                        if (TextUtils.isEmpty(emailAddress)) {
                            "未设置邮箱，无法导出".show(context)
                            return
                        }
                        if (dataBeans.size == 0) {
                            "无打卡记录，无法导出".show(context)
                            return
                        }
                        AlertControlDialog.Builder()
                            .setContext(context)
                            .setTitle("温馨提示")
                            .setMessage("导出到$emailAddress？")
                            .setNegativeButton("取消")
                            .setPositiveButton("确定")
                            .setOnDialogButtonClickListener(object :
                                AlertControlDialog.OnDialogButtonClickListener {
                                override fun onConfirmClick() {
                                    //导出Excel
                                    exportToEmail(dataBeans)
                                }

                                override fun onCancelClick() {}
                            }).build().show()
                    }
                }
            }
        })
    }

    override fun initEvent() {
        binding.leftBackView.setOnClickListener { finish() }

        binding.refreshLayout.setOnRefreshListener { refreshLayout ->
            isRefresh = true
            object : CountDownTimer(1000, 500) {
                override fun onTick(millisUntilFinished: Long) {}
                override fun onFinish() {
                    isRefresh = false
                    dataBeans.clear()
                    offset = 0
                    dataBeans = queryHistoryRecord()
                    refreshLayout.finishRefresh()
                    weakReferenceHandler.sendEmptyMessage(2022021403)
                }
            }.start()
        }
        binding.refreshLayout.setOnLoadMoreListener { refreshLayout ->
            isLoadMore = true
            object : CountDownTimer(1000, 500) {
                override fun onTick(millisUntilFinished: Long) {}
                override fun onFinish() {
                    isLoadMore = false
                    offset++
                    dataBeans.addAll(queryHistoryRecord())
                    refreshLayout.finishLoadMore()
                    weakReferenceHandler.sendEmptyMessage(2022021403)
                }
            }.start()
        }
    }

    override fun observeRequestState() {

    }

    private val callback = Handler.Callback { msg: Message ->
        if (msg.what == 2022021403) {
            if (isRefresh || isLoadMore) {
                historyAdapter.notifyDataSetChanged()
            } else { //首次加载数据
                if (dataBeans.size == 0) {
                    binding.emptyView.visibility = View.VISIBLE
                } else {
                    binding.emptyView.visibility = View.GONE
                    historyAdapter = object :
                        NormalRecyclerAdapter<HistoryRecordBean>(
                            R.layout.item_history_rv_l, dataBeans
                        ) {
                        override fun convertView(
                            viewHolder: ViewHolder, position: Int, item: HistoryRecordBean
                        ) {
                            val message: String = item.message
                            if (!message.contains("成功")) {
                                viewHolder.setImageResource(
                                    R.id.tagView, R.drawable.bg_textview_error
                                )
                            }
                            viewHolder.setText(R.id.noticeMessageView, message)
                                .setText(R.id.noticeDateView, item.date)
                        }
                    }
                    binding.historyRecordView.addItemDecoration(
                        RecyclerViewItemDivider(1, Color.LTGRAY)
                    )
                    binding.historyRecordView.adapter = historyAdapter
                }
            }
        }
        true
    }

    private fun queryHistoryRecord(): MutableList<HistoryRecordBean> {
        return historyRecordBeanDao.queryBuilder()
            .orderDesc(HistoryRecordBeanDao.Properties.Date)
            .offset(offset * 15).limit(15).list()
    }

    private fun exportToEmail(historyBeans: List<HistoryRecordBean>) {
        //{"date":"2020-04-15","message":"考勤打卡:11:42 下班打卡 早退","uuid":"26btND0uLqU"},{"date":"2020-04-15","message":"考勤打卡:16:32 下班打卡 早退","uuid":"UTWQJzCfTl9"}
        val dir = File(this.getExternalFilesDir(Environment.DIRECTORY_DOCUMENTS), "DingRecord")
        if (!dir.exists()) {
            dir.mkdir()
        }

        val fileName = "$dir/打卡记录表.xls"
        ExcelUtils.initExcel(fileName, excelTitle)
        writeObjToExcel(historyBeans, fileName)
    }
}